FGEGraphics/UISystem/UIInputLabel.cs

```diff
@@ -0,0 +1,469 @@
+// If neither of these are available, assume that neither you nor anyone other than the copyright holder
+// hold any right or permission to use this software until such time as the official license is identified.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using FGECore.CoreSystems;
+using FGECore.MathHelpers;
+using FGEGraphics.ClientSystem;
+using FGEGraphics.GraphicsHelpers;
+using FGEGraphics.GraphicsHelpers.FontSets;
+using FGEGraphics.UISystem.InputSystems;
+using OpenTK.Mathematics;
+using OpenTK.Windowing.GraphicsLibraryFramework;
+
+namespace FGEGraphics.UISystem;
+
+/// <summary>Represents an editable text area.</summary>
+// TODO: Text alignment
+// TODO: Horizontal scrolling within bounds, with option
+// TODO: Vertical scrolling when bounds exceeded
+// TODO: Cap text length
+public class UIInputLabel : UIClickableElement
+{
+    /// <summary>An enumeration of <see cref="EditText(EditType, string, string, Action)"/> operations.</summary>
+    public enum EditType
+    {
+        /// <summary>Replaces the space between the indices with the diff.</summary>
+        Add,
+        /// <summary>
+        /// If the indices are not equal, deletes the selection. Otherwise, deletes the character preceding the left index.
+        /// The deleted content becomes the diff.
+        /// </summary>
+        Delete,
+        /// <summary>Submits the 'result' text on close (without a diff).</summary>
+        Submit
+    }
+
+    /// <summary>The text to display when the input is empty.</summary>
+    public UIElementText Info;
+
+    /// <summary>The UI style of normal input content.</summary>
+    public UIElementStyle InputStyle;
+
+    /// <summary>The UI style of highlighted input content.</summary>
+    public UIElementStyle HighlightStyle;
+
+    /// <summary>Whether the input label is currently selected.</summary>
+    public bool Selected = false; // TODO: Provide a UIElement-native solution for this
+
+    /// <summary>Data internal to a <see cref="UIInputLabel"/> instance.</summary>
+    public InternalData Internal = new();
+
+    /// <summary>Fired when the text is edited by the user.</summary>
+    public Action<string> TextEdited;
+
+    /// <summary>Fired when the user submits the text content.</summary>
+    public Action<string> TextSubmitted;
+
+    /// <summary>Fired when the user de-selects the input label.</summary>
+    public Action Closed;
+    
+    /// <summary>Gets or sets the input text content.</summary>
+    public string TextContent
+    {
+        get => Internal.TextContent;
+        set
+        {
+            Internal.SetTextContent(value);
+            UpdateText();
+        }
+    }
+
+    /// <summary>The current number of input text lines.</summary>
+    public int Lines => Internal.TextChain.Sum(piece => piece.Text.Lines.Length);
```

Comment:
Is this correct? Can't left/between/right all be one line, ie this value will be `3` but should actually be `1`?